transport/grpc.go

```diff
@@ -0,0 +1,186 @@
+	"go.uber.org/yarpc/peer/roundrobin"
+	"go.uber.org/yarpc/transport/grpc"
+	"golang.org/x/net/context"
+)
+
+var (
+	errGRPCNoAddresses = errors.New("must specify at least one grpc address")
+	errGRPCNoTracer    = errors.New("must specify grpc tracer")
+	errGRPCNoCaller    = errors.New("must specify grpc caller")
+	errGRPCNoService   = errors.New("must specify grpc service")
+	errGRPCNoEncoding  = errors.New("must specify grpc encoding")
+	errGRPCNoProcedure = errors.New("must specify grpc procedure")
+)
+
+// GRPCOptions are used to create a GRPC transport.
+type GRPCOptions struct {
+	Addresses       []string
+	Tracer          opentracing.Tracer
+	Caller          string
+	Encoding        string
+	RoutingKey      string
+	RoutingDelegate string
+}
+
+// NewGRPC returns a transport that calls a GRPC service.
+func NewGRPC(options GRPCOptions) (TransportCloser, error) {
+	return newGRPC(options)
+}
+
+type grpcTransport struct {
+	Transport       transport.Transport
+	Outbound        transport.UnaryOutbound
+	Caller          string
+	Encoding        string
+	RoutingKey      string
+	RoutingDelegate string
+	tracer          opentracing.Tracer
+}
+
+func newGRPC(options GRPCOptions) (*grpcTransport, error) {
+	if len(options.Addresses) == 0 {
+		return nil, errGRPCNoAddresses
+	}
+	if options.Tracer == nil {
+		return nil, errGRPCNoTracer
+	}
+	if options.Caller == "" {
+		return nil, errGRPCNoCaller
+	}
+	if options.Encoding == "" {
+		return nil, errGRPCNoEncoding
+	}
+
+	transport := grpc.NewTransport(grpc.Tracer(options.Tracer))
+	outbound := transport.NewOutbound(peer.Bind(roundrobin.New(transport), peer.BindPeers(peersToIdentifiers(options.Addresses))))
+	if err := transport.Start(); err != nil {
+		return nil, err
+	}
+	if err := outbound.Start(); err != nil {
+		_ = transport.Stop()
+		return nil, err
+	}
+
+	return &grpcTransport{
+		Transport:       transport,
+		Outbound:        outbound,
+		Caller:          options.Caller,
+		Encoding:        options.Encoding,
+		RoutingKey:      options.RoutingKey,
+		RoutingDelegate: options.RoutingDelegate,
+		tracer:          options.Tracer,
+	}, nil
+}
+
+func (t *grpcTransport) Tracer() opentracing.Tracer {
+	return t.tracer
+}
+
+func (t *grpcTransport) Protocol() Protocol {
+	return GRPC
+}
+
+func (t *grpcTransport) Call(ctx context.Context, request *Request) (*Response, error) {
+	if request.TargetService == "" {
+		return nil, errGRPCNoService
+	}
+	if request.Method == "" {
+		return nil, errGRPCNoProcedure
+	}
+
+	ctx, cancel := requestContextWithTimeout(ctx, request)
+	defer cancel()
+	transportResponse, err := t.Outbound.Call(ctx, t.requestToTransportRequest(request))
+	if err != nil {
+		return nil, err
+	}
+	return transportResponseToResponse(transportResponse)
+}
+
+func (t *grpcTransport) Close() error {
+	return multierr.Combine(t.Transport.Stop(), t.Outbound.Stop())
+}
+
+func (t *grpcTransport) requestToTransportRequest(request *Request) *transport.Request {
```

Comment:
nit: maybe `requestToYARPCRequest`, since we're already in the transport package so saying `ToTransportRequest` is a little unclear